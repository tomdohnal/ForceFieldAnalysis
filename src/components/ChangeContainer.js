import React from 'react';
import { createComponent } from 'react-fela';

import { Box, Text } from './common';

const ChangeContainer = () => (
  <Box padding="20px 16px" color="#268bd2" height="100%">
    <Text>fsfadsfa</Text>
  </Box>
);

export default ChangeContainer;
