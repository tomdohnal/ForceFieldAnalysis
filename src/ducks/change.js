// @flow

export const SET_CHANGE_NAME: string = 'CHANGE/SET_CHANGE_NAME';
export const SET_CHANGE_DESCRIPTION: string = 'CHANGE/SET_CHANGE_DESCRIPTION';

export type Change = {
  name?: string,
  description?: string,
}

type ChangeAction = {
  type: typeof SET_CHANGE_NAME | typeof SET_CHANGE_DESCRIPTION,
  payload: Change,
}

export default (state: Change = {}, action: ChangeAction): Change => {
  switch (action.type) {
  case SET_CHANGE_NAME:
    return { ...state, name: action.payload.name };
  case SET_CHANGE_DESCRIPTION:
    return { ...state, description: action.payload.description };
  default:
    return state;
  }
};

export const setChangeName = (name: string): ChangeAction => ({
  type: SET_CHANGE_NAME,
  payload: {
    name,
  },
});

export const setChangeDescription = (description: string): ChangeAction => ({
  type: SET_CHANGE_DESCRIPTION,
  payload: {
    description,
  },
});
