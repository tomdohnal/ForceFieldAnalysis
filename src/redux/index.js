// @flow
import { combineReducers, createStore } from 'redux';
import _ from 'lodash';

import { loadState, saveState } from './localStorage';

import forcesReducer, { type Forces } from '../ducks/forces';
import changeReducer, { type Change } from '../ducks/change';

export type ReduxState = {
  forces: Forces,
  change: Change,
}

export const RESET_STATE = 'RESET_STATE';

const appReducer = combineReducers({
  forces: forcesReducer,
  change: changeReducer,
});

const rootReducer = (state, action) => {
  if (action.type === RESET_STATE) {
    return appReducer({}, action);
  }

  return appReducer(state, action);
};

const persistedState = loadState();

export const store = createStore(rootReducer, persistedState);

store.subscribe(_.throttle(() => {
  saveState(store.getState());
}, 1000));
