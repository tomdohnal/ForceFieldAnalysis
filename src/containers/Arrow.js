// @flow
import React, { Component } from 'react';
import { Form, Input, Button, Icon, Divider } from 'semantic-ui-react';
import EditIcon from 'react-icons/lib/ti/edit';
import { connect } from 'react-redux';

import { MAX_STRENGTH, MIN_STRENGTH } from '../constants/index';
import { COLORS } from '../constants/styles';
import { type Force, deleteForce, setForceName, increaseForce, decreaseForce } from '../ducks/forces';
import { Text, Box } from '../components/common/index';
import { Rectangle, Triangle } from '../components/svg/index';
import TextBubble from '../components/TextBubble';

type Props = {
  color: string,
  force: Force,
  deleteForce: (id: string) => void,
  setForceName: (id: string, name: string) => void,
  increaseForce: (id: string) => void,
  decreaseForce: (id: string) => void,
}

type State = {
  editArrow: boolean,
}

class Arrow extends Component<Props, State> {
  state = {
    editArrow: false,
  };

  onDeleteArrowIconClick = () => {
    this.props.deleteForce(this.props.force.id);
  };

  onIncreaseArrowIconClick = () => {
    if (this.props.force.strength !== MAX_STRENGTH) {
      this.props.increaseForce(this.props.force.id);
    }
  };

  onDecreaseArrowIconClick = () => {
    if (this.props.force.strength !== MIN_STRENGTH) {
      this.props.decreaseForce(this.props.force.id);
    }
  };

  onEditArrowInputChange = (event, data) => {
    this.props.setForceName(this.props.force.id, data.value);
  };

  onEditArrowNameClick = () => {
    this.setState({ editArrow: true });
  };

  onEditArrowNameFormSubmit = () => {
    this.setState({ editArrow: false });
  };

  render() {
    const { force: { driving, strength, name }, color } = this.props;

    const arrowUnitWidth = 50;
    const arrowWidth = strength * arrowUnitWidth;
    const arrowHeight = 80;
    const rectangleHeight = arrowHeight - 40;
    const triangleWidth = 36;

    return (
      <Box>
        <TextBubble color={color} left={!driving}>
          {this.state.editArrow ?
            <Form onSubmit={this.onEditArrowNameFormSubmit}>
              <Input autoFocus size="small" value={name} onChange={this.onEditArrowInputChange} action>
                <input style={{ width: 'inherit' }} />
                <Button size="small">Save changes</Button>
              </Input>
            </Form>
            :
            <Box>
              <Text fontSize="15px">{name}</Text>
              <EditIcon
                size="18"
                style={{ verticalAlign: 'top', marginLeft: '4px', cursor: 'pointer' }}
                onClick={this.onEditArrowNameClick}
              />
            </Box>
          }
        </TextBubble>
        <Box position="relative">
          <Icon
            name="delete"
            size="big"
            style={{
              position: 'absolute',
              top: 'calc(50% - 14px)',
              color: COLORS.RED,
              cursor: 'pointer',
              [driving ? 'right' : 'left']: `${arrowWidth}px`,
            }}
            onClick={this.onDeleteArrowIconClick}
          />
          <svg height={arrowHeight} width={arrowWidth} viewBox={`0 0 ${arrowWidth} ${arrowHeight}`}>
            <Rectangle
              width={arrowWidth - triangleWidth}
              height={rectangleHeight}
              color={color}
              x={!driving ? triangleWidth : 0}
              y={(arrowHeight - rectangleHeight) / 2}
            />
            <Triangle
              width={triangleWidth}
              height={arrowHeight}
              x={`${!driving ? triangleWidth : arrowWidth - triangleWidth},0`}
              y={`${!driving ? triangleWidth : arrowWidth - triangleWidth},${arrowHeight}`}
              z={`${!driving ? 0 : arrowWidth},${arrowHeight / 2}`}
              color={color}
            />
          </svg>
          <Box
            position="absolute"
            top={`${arrowHeight - ((arrowHeight - rectangleHeight) / 2)}px`}
            right={driving && `${triangleWidth}px`}
            left={!driving && `${triangleWidth}px`}
          >
            <Icon
              name="plus"
              size="large"
              disabled={strength === MAX_STRENGTH}
              style={{ cursor: `${strength < MAX_STRENGTH ? 'pointer' : 'initial'}` }}
              onClick={this.onIncreaseArrowIconClick}
            />
            <Text fontSize="18px">{strength}</Text>
            <Icon
              name="minus"
              size="large"
              disabled={strength === MIN_STRENGTH}
              style={{ cursor: `${strength > MIN_STRENGTH ? 'pointer' : 'initial'}` }}
              onClick={this.onDecreaseArrowIconClick}
            />
          </Box>
        </Box>
        <Divider hidden />
      </Box>
    );
  }
}

const actionCreators = { deleteForce, setForceName, increaseForce, decreaseForce };

export default connect(undefined, actionCreators)(Arrow);
