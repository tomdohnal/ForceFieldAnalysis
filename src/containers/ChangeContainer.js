// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Input, Button } from 'semantic-ui-react';

import { COLORS } from '../constants/styles';
import type { ReduxState } from '../redux/index';
import { setChangeName, setChangeDescription, type Change } from '../ducks/change';
import { Box, Text } from '../components/common/index';

type Props = {
  change: Change,
  onNameChange: (name: string) => void,
  onDescriptionChange: (description: string) => void,
};

type State = {
  editChangeName: boolean,
  editChangeDescription: boolean,
};

class ChangeContainer extends Component<Props, State> {
  state = {
    editChangeName: !this.props.change.name,
    editChangeDescription: !this.props.change.description,
  };

   onButtonClick = editChangeBoolean => (): void => {
     this.setState({ [editChangeBoolean]: false });
   };

   renderForm(onChange: (value: string) => void, onButtonClick: (value: string) => void) {
     return (
       <Box>
         <Input onChange={onChange} />
         <Button onClick={onButtonClick} />
       </Box>
     );
   }

   render() {
     const { change: { name, description }, onNameChange, onDescriptionChange } = this.props;

     return (
       <Box padding="20px 16px" color={COLORS.BLUE} height="100%">
         {this.state.editChangeName ?
           this.renderForm(onNameChange, this.onButtonClick('editChangeName')) :
           <Text>{name}</Text>
         }
         {this.state.editChangeDescription ?
           this.renderForm(onDescriptionChange, this.onButtonClick('editChangeDescription')) :
           <Text>{description}</Text>
         }
       </Box>
     );
   }
}

const mapStateToProps = (state: ReduxState) => ({
  change: state.change,
});

const mapDispatchToProps = dispatch => ({
  onNameChange: (event, data) => (dispatch(setChangeName(data.value))),
  onDescriptionChange: (event, data) => (dispatch(setChangeDescription(data.value))),
});

export default connect(mapStateToProps, mapDispatchToProps)(ChangeContainer);
