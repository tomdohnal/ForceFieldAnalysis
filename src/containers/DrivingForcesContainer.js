// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Input, Button } from 'semantic-ui-react';

import DrivingForceArrow from '../components/DrivingForceArrow';
import { addDrivingForce, type Forces } from '../ducks/forces';

type State = {
  newDrivingForceName: string,
  newDrivingForceError: boolean,
  newDrivingForceErrorText: string,
};

type Props = {
  forces: Forces,
  addDrivingForce: (name: string, strength: number) => void,
};

class DrivingForcesContainer extends Component<Props, State> {
  state = {
    newDrivingForceName: '',
    newDrivingForceError: false,
    newDrivingForceErrorText: '',
  };

  onNewDrivingForceInputChange = (e) => {
    this.setState({
      newDrivingForceName: e.target.value,
      newDrivingForceError: false,
      newDrivingForceErrorText: '',
    });
  };

  onNewDrivingForceButtonClick = () => {
    if (!this.state.newDrivingForceName) {
      this.setState({
        newDrivingForceError: true,
        newDrivingForceErrorText: 'Enter the description for the force.',
      });
    } else {
      this.setState({
        newDrivingForceName: '',
      });

      this.props.addDrivingForce(this.state.newDrivingForceName, 2);
    }
  };

  renderForces() {
    return this.props.forces.map(force => (
      <DrivingForceArrow strength={force.strength} key={force.id} />));
  }

  render() {
    return (
      <div>
        {this.renderForces()}
        <Input
          onChange={this.onNewDrivingForceInputChange}
          value={this.state.newDrivingForceName}
          error={this.state.newDrivingForceError}
        />
        <Button onClick={this.onNewDrivingForceButtonClick}>New driving force</Button>
        <span>{this.state.newDrivingForceErrorText}</span>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  forces: state.forces.filter(force => force.driving === true),
});

export default connect(mapStateToProps, { addDrivingForce })(DrivingForcesContainer);
