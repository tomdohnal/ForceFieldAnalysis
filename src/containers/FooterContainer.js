// @flow
import React from 'react';
import { connect } from 'react-redux';
import { Grid } from 'semantic-ui-react';

import { type Force } from '../ducks/forces';
import { type ReduxState } from '../redux';
import Footer from '../components/Footer';
import Text from '../components/Text';

type Props = {
  drivingForcesCount: number,
  hinderingForcesCount: number,
}

const FooterContainer = ({
  drivingForcesCount, hinderingForcesCount,
}: Props) => {
  let centerText: string = '';
  if (drivingForcesCount === 0 && hinderingForcesCount === 0) {
    centerText = 'Create some forces and start the analysis!';
  } else if (drivingForcesCount === hinderingForcesCount) {
    centerText = 'It\'s a tie!';
  } else if (drivingForcesCount > hinderingForcesCount) {
    centerText = 'Let\'s make the change happen!';
  } else {
    centerText = 'No way! Status quo wins...';
  }

  return (
    <Footer>
      <Grid>
        <Grid.Column width={6}>
          <Text fontSize="54px" color="#FFF">Driving forces: {drivingForcesCount}</Text>
        </Grid.Column>
        <Grid.Column width={4} textAlign="center">
          <Text fontSize="24px" color="#FFF">{centerText}</Text>
        </Grid.Column>
        <Grid.Column width={6} textAlign="right">
          <Text fontSize="54px" color="#FFF">Hindering forces: {hinderingForcesCount}</Text>
        </Grid.Column>
      </Grid>
    </Footer>
  );
};


const mapStateToProps = (state: ReduxState) => {
  const countForceTotal = (driving: boolean) => (total: number, force: Force) => {
    if (force.driving === driving) {
      return total + force.strength;
    }

    return total;
  };

  return {
    drivingForcesCount: state.forces.length ?
      state.forces.reduce(countForceTotal(true), 0) : 0,
    hinderingForcesCount: state.forces.length ?
      state.forces.reduce(countForceTotal(false), 0) : 0,
  };
};

export default connect(mapStateToProps)(FooterContainer);
