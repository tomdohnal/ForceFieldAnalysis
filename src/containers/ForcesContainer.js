// @flow
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Form, Input, Button, Select, Label } from 'semantic-ui-react';

import DrivingForceArrow from '../components/DrivingForceArrow';
import HinderingForceArrow from '../components/HinderingForceArrow';
import { Box } from '../components/common';
import { addDrivingForce, addHinderingForce, type Forces } from '../ducks/forces';
import { type ReduxState } from '../redux';

type State = {
  newForceName: string,
  newForceError: boolean,
  newForceErrorText: string,
  newForceStrength: number,
};

type Props = {
  driving: boolean,
  submitButtonText: string,
  forces: Forces,
  addForce: (name: string, strength: number) => void,
};

class ForcesContainer extends Component<Props, State> {
  defaultState = {
    newForceName: '',
    newForceError: false,
    newForceErrorText: '',
    newForceStrength: 1,
  };

  state = this.defaultState;

  onNewForceInputChange = (e, data) => {
    this.setState({
      newForceName: data.value,
      newForceError: false,
      newForceErrorText: '',
    });
  };

  onNewForceStrengthDropdownChange = (e, data) => {
    this.setState({
      newForceStrength: data.value,
    });
  };

  onNewForceButtonClick = () => {
    if (!this.state.newForceName) {
      this.setState({
        newForceError: true,
        newForceErrorText: 'Enter the description for the force.',
      });
    } else {
      this.props.addForce(this.state.newForceName, this.state.newForceStrength);

      this.setState(this.defaultState);
    }
  };

  getStrengthOptions() {
    let options = [];

    for (let i = 1; i <= 8; i += 1) {
      options = [...options, { key: i, text: i, value: i }];
    }

    return options;
  }

  renderForces() {
    return this.props.forces.map((force) => {
      if (this.props.driving) {
        return <DrivingForceArrow strength={force.strength} key={force.id} />;
      }
      return <HinderingForceArrow strength={force.strength} key={force.id} />;
    });
  }

  render() {
    return (
      <Box padding="0 16px">
        {this.renderForces()}
        <Form>
          <Form.Group inline>
            <Form.Field>
              <label>Force name:</label>
              {this.state.newForceError ? <Label basic color="red" pointing="below">{this.state.newForceErrorText}</Label> : ''}
              <Input
                onChange={this.onNewForceInputChange}
                value={this.state.newForceName}
                error={this.state.newForceError}
              />
            </Form.Field>
            <Form.Field>
              <label>Force strength</label>
              <Select
                compact
                labeled
                onChange={this.onNewForceStrengthDropdownChange}
                options={this.getStrengthOptions()}
                value={this.state.newForceStrength}
              />
            </Form.Field>
          </Form.Group>
          <Form.Field>
            <Button onClick={this.onNewForceButtonClick}>{this.props.submitButtonText}</Button>
          </Form.Field>
        </Form>
      </Box>
    );
  }
}

const mapStateToProps = (state: ReduxState, ownProps: Props) => ({
  forces: state.forces.filter(force => force.driving === ownProps.driving),
});

const mapDispatchToProps = (dispatch, ownProps: Props) => ({
  addForce: (name: string, strength: number) => (
    ownProps.driving ?
      dispatch(addDrivingForce(name, strength)) :
      dispatch(addHinderingForce(name, strength))
  ),
});

export default connect(mapStateToProps, mapDispatchToProps)(ForcesContainer);
