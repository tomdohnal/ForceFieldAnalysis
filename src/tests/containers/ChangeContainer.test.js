import React from 'react';
import { shallow } from 'enzyme';

import { store } from '../../redux/index';
import { setChangeName, setChangeDescription } from '../../ducks/change';
import { ChangeContainer, mapStateToProps } from '../../containers/ChangeContainer';

describe('ChangeContainer', () => {
  describe('rendering', () => {
    it('renders correctly with no change name and description', () => {
      const tree = shallow(<ChangeContainer
        change={{
          name: '',
          description: '',
        }}
      />);

      expect(tree).toMatchSnapshot();
    });

    it('renders correctly with change name and description', () => {
      const tree = shallow(
        <ChangeContainer
          change={{
            name: 'change name',
            description: 'change description',
          }}
        />,
      );
      expect(tree).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    let ShallowChangeContainer;

    // in order for the component's state to reinitialize before every test,
    // we create a new component before each test
    beforeEach(() => {
      ShallowChangeContainer = shallow(
        <ChangeContainer
          change={{
            name: 'change name',
            description: 'change description',
          }}
          setChangeName={jest.fn()}
          setChangeDescription={jest.fn()}
        />,
      );
    });

    it('submits name form', () => {
      ShallowChangeContainer.instance().onNameFormSubmit();

      expect(ShallowChangeContainer.state('editChangeName')).toBeFalsy();
    });

    it('enables you to edit the name', () => {
      ShallowChangeContainer.instance().onEditNameClick();

      expect(ShallowChangeContainer.state('editChangeName')).toBeTruthy();
    });

    it('submits description form', () => {
      ShallowChangeContainer.instance().onDescriptionFormSubmit();

      expect(ShallowChangeContainer.state('editChangeDescription')).toBeFalsy();
    });

    it('closes the description form on No Description Button click', () => {
      const mockEvent = { preventDefault() {} };

      ShallowChangeContainer.instance().onNoDescriptionButtonClick(mockEvent);

      expect(ShallowChangeContainer.state('editChangeDescription')).toBeFalsy();
    });

    it('clears the description from on No Description Button click', () => {
      const mockEvent = { preventDefault() {} };

      ShallowChangeContainer.instance().onNoDescriptionButtonClick(mockEvent);

      expect(ShallowChangeContainer.instance().props.setChangeDescription).toBeCalledWith('');
    });

    it('enables you to edit the description', () => {
      ShallowChangeContainer.instance().onEditDescriptionClick();

      expect(ShallowChangeContainer.state('editChangeDescription')).toBeTruthy();
    });

    it('calls the setChangeName actions creator when onNameChange is called', () => {
      ShallowChangeContainer.instance().onNameChange({}, { value: 'change name' });

      expect(ShallowChangeContainer.instance().props.setChangeName).toBeCalledWith('change name');
    });

    it('calls the setChangeDescription actions creator when onDescriptionChange is called', () => {
      ShallowChangeContainer.instance().onDescriptionChange({}, { value: 'change description' });

      expect(ShallowChangeContainer.instance().props.setChangeDescription).toBeCalledWith('change description');
    });
  });

  describe('mapStateToProps', () => {
    it('returns the change', () => {
      store.dispatch(setChangeName('change name'));
      store.dispatch(setChangeDescription('change description'));

      const change = mapStateToProps(store.getState());

      expect(change).toEqual({ change: { name: 'change name', description: 'change description' } });
    });
  });
});
