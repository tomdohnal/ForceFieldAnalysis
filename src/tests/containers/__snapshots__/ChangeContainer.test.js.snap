// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeContainer rendering renders correctly with change name and description 1`] = `
<boxStyles
  backgroundColor="#268bd2"
  height="100%"
  padding="20px 16px"
>
  <boxStyles
    marginTop="16px"
    textAlign="center"
  >
    <text
      color="#fff"
      fontSize="48px"
    >
      change name
    </text>
    <boxStyles
      display="inline"
    >
      <TiEdit
        color="#fff"
        onClick={[Function]}
        size="28"
        style={
          Object {
            "cursor": "pointer",
            "marginLeft": "8px",
            "verticalAlign": "top",
          }
        }
      />
    </boxStyles>
  </boxStyles>
  <boxStyles
    marginTop="16px"
  >
    <text
      color="#fff"
      fontSize="18px"
    >
      change description
    </text>
    <boxStyles
      display="inline"
    >
      <TiEdit
        color="#fff"
        onClick={[Function]}
        size="18"
        style={
          Object {
            "cursor": "pointer",
            "marginLeft": "4px",
            "verticalAlign": "top",
          }
        }
      />
    </boxStyles>
  </boxStyles>
</boxStyles>
`;

exports[`ChangeContainer rendering renders correctly with no change name and description 1`] = `
<boxStyles
  backgroundColor="#268bd2"
  height="100%"
  padding="20px 16px"
>
  <ChangeNameForm
    inputValue=""
    onInputChange={[Function]}
    onSubmit={[Function]}
  />
  <boxStyles
    marginTop="16px"
  >
    <ChangeDescriptionForm
      inputValue=""
      onInputChange={[Function]}
      onNoDescriptionButtonClick={[Function]}
      onSubmit={[Function]}
    />
  </boxStyles>
</boxStyles>
`;
